### PR TITLE
WT-15238 Fix prepared commit and rollback race with reconciliation

### PR DESCRIPTION
**Pull request by quchenhao**
Change the way we handle prepared commit and rollback. It will no longer mark anything to be deleted from the history store. It will just append the history store record to the update chain if necessary and resolve the prepared updates. Reconciliation will handling deleting the updates from the history store and inserting updates to the history store alone. It is doing that based on the new flag WT_UPDATE_HS_MAX_STOP, which means an update is inserted into the history store with a max time point. Reconciliation will need to delete the update with this flag from the history store if we write a committed update to the data store. In addition, if we select an update has been inserted to the history store to write to the data store, we also need to delete that record from the history store first.

Fix an issue that update restore eviction for prepared updates need to retain the oldest full value to resolve modifies. Add a test for that.

Fix a race that checkpoint may see an aborted update in prepare locked state.

Fix a case that prepare insert rollback is lost. The scenario is like this:
* We have a prepared insert.
* We rollback the prepared insert.
* Eviction writes it as prepared to disk.
* We restore back the aborted prepared.
* The rollback timestamp becomes stable.
* We ignore it and still keep the existing on-disk prepared as it is.
* Another reader reads the page and hits a prepared conflict.

### CODE REVIEW

File: src/btree/row_modify.c [hunk 1]
```diff
@@ -429,7 +429,7 @@ __wt_update_obsolete_check(
             first = NULL;
 
         /* Cannot truncate the updates if we need to remove the updates from the history store. */
-        if (F_ISSET(upd, WT_UPDATE_TO_DELETE_FROM_HS))
+        if (F_ISSET(upd, WT_UPDATE_HS_MAX_STOP))
```
Comment (kommiharibabu) [inline src/btree/row_modify.c:437]: We do use this flag to remove the update from HS even if it has the stop also because the data store is rolled back. Using the flag HS_MAX_STOP may be confusing?

Comment (quchenhao) [reply src/btree/row_modify.c:437]: We no longer mark them to be deleted. Reconciliation itself will decide which to delete which not so I introduced this flag to distinguish with a proper hs record with a valid stop time point

Comment (kommiharibabu) [reply src/btree/row_modify.c:437]: ok.

File: src/reconcile/rec_visibility.c [hunk 2]
```diff
@@ -265,64 +256,54 @@ static int
...
-                /*
-                 * It must be the operation on this prepared update that triggered us to delete the
-                 * update from the history store. If it is the same prepared update we are writing
-                 * to disk, don't delete the update from the history store now. We need to wait the
-                 * prepared update to become stable.
-                 */
-                if (txnid == upd_select->tw.start_txn)
-                    break;
-            }
+        if (!seen_committed && upd_select->tw.start_ts != txnid)
```
Comment (tdlinh2712) [inline src/reconcile/rec_visibility.c]: this should be `upd_select->tw.start_txn != txnid`?

Comment (quchenhao) [reply src/reconcile/rec_visibility.c]: good catch. thank you

File: src/txn/txn.c [hunk 3]
```diff
@@ -890,11 +885,9 @@ __txn_prepare_rollback_restore_hs_update(
...
         tombstone->txnid = hs_tw->stop_txn;
         tombstone->next = upd;
-        /*
-         * Set the flag to indicate that this update has been restored from history store for the
-         * rollback of a prepared transaction.
-         */
-        F_SET(tombstone, WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+        /* Set the flag to indicate that this update has been restored from history store. */
+        F_SET(tombstone, WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_HS);
+        F_SET(upd, WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_HS);
```
Comment (tdlinh2712) [inline src/txn/txn.c:890]: in line 90 we will assign `upd = tombstone`, do we need to set the flag for upd here?

Comment (quchenhao) [reply src/txn/txn.c:890]: yes, the code can be simplified

Comment (quchenhao) [reply src/txn/txn.c:890]: I would prefer doing this way. It is clear we set flags for both updates.

File: src/reconcile/rec_visibility.c [hunk 4]
```diff
@@ -265,64 +256,54 @@ static int
...
-            if (prepare_state != WT_PREPARE_INIT)
-                triggering_prepare_upd = delete_upd;
+    if (F_ISSET(upd_select->upd, WT_UPDATE_HS | WT_UPDATE_HS_MAX_STOP)) {
+        if (upd_select->tombstone == NULL)
+            tombstone = NULL;
+        else if (F_ISSET(upd_select->tombstone, WT_UPDATE_HS))
+            tombstone = upd_select->tombstone;
+        else {
+            tombstone = NULL;
+            WT_ASSERT(session, !F_ISSET(upd_select->tombstone, WT_UPDATE_HS_MAX_STOP));
```
Comment (tdlinh2712) [inline src/reconcile/rec_visibility.c:281]: I don't really understand this assert. Why can't a tombstone without `WT_UPDATE_HS` have `WT_UPDATE_HS_MAX_STOP`? Can we add a comment to clarify this?

File: src/reconcile/rec_visibility.c [hunk 5]
```diff
@@ -265,64 +256,54 @@ static int
...
-                 * It must be the operation on this prepared update that triggered us to delete the
-                 * update from the history store. If it is the same prepared update we are writing
-                 * to disk, don't delete the update from the history store now. We need to wait the
-                 * prepared update to become stable.
-                 */
-                if (txnid == upd_select->tw.start_txn)
-                    break;
-            }
+        if (!seen_committed && upd_select->tw.start_txn != txnid)
+            seen_committed = true;
```
Comment (tdlinh2712) [inline src/reconcile/rec_visibility.c:300]: I think the reason why`seen_committed` is set here is not very straightforward. Can we add a comment why `seen_committed` is set to true here? something like
```
If we find an update from a different transaction than the update being written to disk, it means that this update must already been committed, so from this point onwards it's safe to delete all updates with `WT_UPDATE_HS_MAX_STOP` flag
```

File: src/reconcile/rec_visibility.c [hunk 6]
```diff
@@ -265,64 +256,54 @@ static int
...
+            /*
+             * Don't delete it from the history store if we again write a prepared update to disk
+             * and there is nothing in between the prepared update and the update with the max stop
+             * point in the history store.
+             */
+            if (!seen_committed)
                 break;
-            }
+            WT_RET(__rec_delete_hs_upd_save(session, r, ins, rip, delete_upd, NULL));
+            break;
```
Comment (tdlinh2712) [inline src/reconcile/rec_visibility.c:318]: The comment says "Delete all the history store record with a max stop timestamp" but if we break here, only one record with max stop timestamp will be deleted?

I also have another question for the break in line `300`. Consider this scenario:
```
[upd_select (prepared)] -> [hs_entry1 (WT_UPDATE_HS_MAX_STOP)] -> [committed_txn] -> [hs_entry2 (WT_UPDATE_HS_MAX_STOP)]
```
Since `upd_select` is written as prepared, `seen_committed` is false and we will break early at `hs_entry1` and won't delete `hs_entry2`. Is this expected?

Comment (quchenhao) [reply src/reconcile/rec_visibility.c:318]: This case is not possible. There has only one maximum record with max stop timestamp in the history store for a key. When a new record with max time point is inserted to the hs, any previous record with max hs must have been deleted.

File: src/reconcile/rec_visibility.c (hunk 4 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:281]: done

File: src/reconcile/rec_visibility.c (hunk 5 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:300]: done

File: src/reconcile/rec_visibility.c (hunk 6 above)
Comment (quchenhao) [reply src/reconcile/rec_visibility.c:318]: done

Comment (tdlinh2712) [reply src/reconcile/rec_visibility.c:318]: why is that the case? If no reconciliation is called between two transactions, how is the previous record with max ts be deleted?

Comment (quchenhao) [reply src/reconcile/rec_visibility.c:318]: Writing a prepared to disk needs a reconciliation. At that reconciliation, it will delete the previous max stop time and then insert the new one as max stop. So you cannot have two updates both with max in hs. Anyway you will get an error when you insert the second one with max stop when there is an existing one in hs.

Comment (tdlinh2712) [reply src/reconcile/rec_visibility.c:318]: Ah i see, thank you!

**Review by tdlinh2712 — APPROVED** [review]
lgtm, thanks!

**Comment by quchenhao** [comment]
> How risky is this change? Why?

This change is risky as it changes the way we resolve and reconcile prepared update and impacts both ASC and DSC. However, we need to do this because with the precise checkpoint writing prepared update in checkpoint, we can race with prepared commit and rollback. This change makes reconciliation solely responsible for deleting data with max stop time point or should be restored from the history store to avoid this race.

> What tests are you adding, changing or relying on? Why?

This change is necessary for enabling stress testing for preserve prepared. It also has two python tests for the two edge cases it fixes.

> What, if anything, are you concerned about that you'd like the reviewer to focus on?

The logic of the code and any potential race conditions not considered.

File: src/btree/bt_split.c [hunk 7]
```diff
@@ -1526,9 +1526,11 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
             /*
              * If we have written a prepared update, we need to retain the next update that is not a
              * tombstone. Otherwise, we don't have anything to write in the next reconciliation if
-             * the prepared update is reverted.
+             * the prepared update is reverted. If the next value update is a modify, we need to
+             * retain all the older updates until a full value is found.
              */
             if (WT_TIME_WINDOW_HAS_START_PREPARE(&supd->tw)) {
+                bool find_next_value = false;
```
Comment (kommiharibabu) [inline src/btree/bt_split.c]: I didn't clearly understand the use of this `find_next_value` flag. Doesn't adding `if (WT_UPDATE_DATA_VALUE(tmp))` code before `break` solve the requirement?

File: test/suite/test_prepare41.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        evict_cursor = session_evict.open_cursor(self.uri, None, None)
+        for i in range(1, 21):  # Evict committed data pages
+            evict_cursor.set_key(i)
+            self.assertEqual(evict_cursor.search(), 0)
+            evict_cursor.reset()
+        evict_cursor.close()
+        session_evict.rollback_transaction()
+
+        self.conn.set_timestamp('stable_timestamp=' + self.timestamp_str(40))
+        # We should not crash
```
Comment (kommiharibabu) [inline test/suite/test_prepare41.py]: This comment needs an update.

File: src/btree/bt_split.c (hunk 7 above)
Comment (quchenhao) [reply src/btree/bt_split.c]: This is to exclude the tombstone. We need a full standard update. I can simplify this to:

```
if (tmp->type == WT_UPDATE_STANDARD)
        break;
```

Comment (kommiharibabu) [reply src/btree/bt_split.c]: Yes. that would be good.

File: test/suite/test_prepare41.py (hunk 8 above)
Comment (quchenhao) [reply test/suite/test_prepare41.py]: done

**Review by kommiharibabu — APPROVED** [review]
LGTM
